File: frontend/src/lib/hash.ts
import { sha256 } from "@noble/hashes/sha256";
import type { SwapDetails } from "./program";

// ============================================================================
// BORSH SERIALIZATION HELPERS
// ============================================================================

/**
 * Convert a bigint to a little-endian Uint8Array (u64)
 */
function bigintToU64LE(value: bigint): Uint8Array {
  const buffer = new ArrayBuffer(8);
  const view = new DataView(buffer);
  view.setBigUint64(0, value, true); // little-endian
  return new Uint8Array(buffer);
}

/**
 * Convert a number to a little-endian Uint8Array (u16)
 */
function numberToU16LE(value: number): Uint8Array {
  const buffer = new ArrayBuffer(2);
  const view = new DataView(buffer);
  view.setUint16(0, value, true); // little-endian
  return new Uint8Array(buffer);
}

// ============================================================================
// SWAP DETAILS HASHING
// ============================================================================

/**
 * Serialize SwapDetails to match on-chain Borsh serialization
 * Layout: amount_in (u64 LE) + min_out (u64 LE) + slippage_bps (u16 LE) + nonce ([u8; 32])
 * Total: 8 + 8 + 2 + 32 = 50 bytes
 */
export function serializeSwapDetails(details: SwapDetails): Uint8Array {
  const amountInBytes = bigintToU64LE(details.amountIn);
  const minOutBytes = bigintToU64LE(details.minOut);
  const slippageBpsBytes = numberToU16LE(details.slippageBps);
  
  // Combine all parts
  const result = new Uint8Array(50);
  result.set(amountInBytes, 0);       // bytes 0-7
  result.set(minOutBytes, 8);         // bytes 8-15
  result.set(slippageBpsBytes, 16);   // bytes 16-17
  result.set(details.nonce, 18);      // bytes 18-49
  
  return result;
}

/**
 * Hash SwapDetails using SHA-256
 * This must match the on-chain hashing exactly
 */
export function hashSwapDetails(details: SwapDetails): Uint8Array {
  const serialized = serializeSwapDetails(details);
  return sha256(serialized);
}

/**
 * Generate a random nonce for replay protection
 */
export function generateNonce(): Uint8Array {
  const nonce = new Uint8Array(32);
  if (typeof window !== "undefined" && window.crypto) {
    window.crypto.getRandomValues(nonce);
  } else {
    // Fallback for Node.js environment
    for (let i = 0; i < 32; i++) {
      nonce[i] = Math.floor(Math.random() * 256);
    }
  }
  return nonce;
}

/**
 * Create SwapDetails and compute its hash
 * Returns both the details and the hash for use in commit/reveal
 */
export function createSwapDetailsWithHash(
  amountIn: bigint,
  minOut: bigint,
  slippageBps: number
): { details: SwapDetails; hash: Uint8Array; nonce: Uint8Array } {
  const nonce = generateNonce();
  
  const details: SwapDetails = {
    amountIn,
    minOut,
    slippageBps,
    nonce,
  };
  
  const hash = hashSwapDetails(details);
  
  return { details, hash, nonce };
}

/**
 * Convert hash bytes to hex string for display
 */
export function hashToHex(hash: Uint8Array): string {
  return Array.from(hash)
    .map(b => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Convert hash to array format for instruction arguments
 */
export function hashToArray(hash: Uint8Array): number[] {
  return Array.from(hash);
}

File: mev-simulation/src/utils/hash.rs
//! Hash Utilities for Commit-Reveal
//!
//! Implements SHA256 hashing that matches the on-chain program's
//! Borsh serialization format.

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

/// Swap details that are hashed for the commit-reveal scheme
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SwapDetails {
    /// Amount of input tokens in lamports
    pub amount_in: u64,
    /// Minimum output amount (slippage protection)
    pub min_out: u64,
    /// Slippage tolerance in basis points
    pub slippage_bps: u16,
    /// Random nonce for replay protection (32 bytes)
    pub nonce: [u8; 32],
}

impl SwapDetails {
    /// Create new swap details with a random nonce
    pub fn new(amount_in: u64, min_out: u64, slippage_bps: u16) -> Self {
        let mut nonce = [0u8; 32];
        rand::Rng::fill(&mut rand::thread_rng(), &mut nonce);
        
        Self {
            amount_in,
            min_out,
            slippage_bps,
            nonce,
        }
    }

    /// Create swap details with a specific nonce
    pub fn with_nonce(amount_in: u64, min_out: u64, slippage_bps: u16, nonce: [u8; 32]) -> Self {
        Self {
            amount_in,
            min_out,
            slippage_bps,
            nonce,
        }
    }

    /// Serialize to bytes matching on-chain Borsh format
    /// Layout: amount_in (u64 LE) + min_out (u64 LE) + slippage_bps (u16 LE) + nonce ([u8; 32])
    /// Total: 8 + 8 + 2 + 32 = 50 bytes
    pub fn serialize(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(50);
        
        // amount_in: u64 little-endian
        bytes.extend_from_slice(&self.amount_in.to_le_bytes());
        
        // min_out: u64 little-endian
        bytes.extend_from_slice(&self.min_out.to_le_bytes());
        
        // slippage_bps: u16 little-endian
        bytes.extend_from_slice(&self.slippage_bps.to_le_bytes());
        
        // nonce: [u8; 32]
        bytes.extend_from_slice(&self.nonce);
        
        bytes
    }
}

/// Hash swap details using SHA256
/// Returns a 32-byte hash matching the on-chain commitment
pub fn hash_swap_details(details: &SwapDetails) -> [u8; 32] {
    let serialized = details.serialize();
    let mut hasher = Sha256::new();
    hasher.update(&serialized);
    let result = hasher.finalize();
    
    let mut hash = [0u8; 32];
    hash.copy_from_slice(&result);
    hash
}

/// Convert hash to hex string for display
pub fn hash_to_hex(hash: &[u8; 32]) -> String {
    hash.iter()
        .map(|b| format!("{:02x}", b))
        .collect()
}

/// Generate a random 32-byte nonce
pub fn generate_nonce() -> [u8; 32] {
    let mut nonce = [0u8; 32];
    rand::Rng::fill(&mut rand::thread_rng(), &mut nonce);
    nonce
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_serialization_length() {
        let details = SwapDetails::new(1_000_000_000, 900_000_000, 100);
        let serialized = details.serialize();
        assert_eq!(serialized.len(), 50);
    }

    #[test]
    fn test_hash_determinism() {
        let nonce = [42u8; 32];
        let details1 = SwapDetails::with_nonce(1_000_000_000, 900_000_000, 100, nonce);
        let details2 = SwapDetails::with_nonce(1_000_000_000, 900_000_000, 100, nonce);
        
        let hash1 = hash_swap_details(&details1);
        let hash2 = hash_swap_details(&details2);
        
        assert_eq!(hash1, hash2);
    }

    #[test]
    fn test_different_inputs_different_hashes() {
        let nonce = [42u8; 32];
        let details1 = SwapDetails::with_nonce(1_000_000_000, 900_000_000, 100, nonce);
        let details2 = SwapDetails::with_nonce(2_000_000_000, 900_000_000, 100, nonce);
        
        let hash1 = hash_swap_details(&details1);
        let hash2 = hash_swap_details(&details2);
        
        assert_ne!(hash1, hash2);
    }
}

